articles/events_executor.md

```diff
@@ -0,0 +1,233 @@
+   This can flow into the "You don't pay for what you don't use" paradigm.
+ - Extensibility: The executor should set a base from which it is straight-forward to add more functionalities without impacting performances or requiring heavy modifications of the ROS2 core libraries.
+ - Ordering of events: The executor should process events in the same order they have occurred.
+ - High precision and accurate timers: The executor should provide mechanisms to ensure that timers are executed with the minimum possible latency.
+ 
+## The Events Executor
+
+In order to address the aforementioned requirements, we designed the so-called `EventsExecutor`.
+The `EventsExecutor` inherits from the base `Executor` class.
+The main difference between this and the other executor implementations is that the `EventsExecutor` does not use the concept of a waitset.
+The `EventsExecutor` is characterized by two main components, i.e. the `EventsQueue` and the `TimersManager`.
+
+### Brief Overview
+
+In order to execute generic ROS 2 entities (i.e. everything except timers), the `EventsExecutor` relies on an event queue.
+Whenever a ROS 2 entity that is associated to the executor has work to do, it will push an event into the executor's queue.
+Then the executor can process these events in a FIFO manner, without need for expensive entities look-ups.
+Processing an event results in different operations depending on the entity that generated it.
+
+Timers are executed in a separate task by a timers manager, where they are kept in a priority queue sorted according to their expiration time. The task then has to monitor only the first element of the queue and can execute its callback as soon as it expires.
+
+![Overview](../img/events_executor/overview.png)
+
+
+### EventsQueue
+
+The `EventsExecutor` requires that all the entities (except ROS 2 timers) as soon as they have some work to do, they push an event into the executor's `EventsQueue`.
+The events are produced in the RMW layer, i.e. where the underlying middleware notifies the ROS 2 entity that there is new work to do, on the other hand, events are executed in the rclcpp layer by the `EventsExecutor`.
+The event data structure that is pushed into the `EventsQueue` must contain all what is needed for the executor to be able to process that particular event.
+The data structure so includes the type of the entity that generated the event and a handle to its corresponding rclcpp object.
+
+Considering the entities currently available in ROS 2, the content of the event data structure can be any of the following:
+ - `ExecutorEventType::SUBSCRIPTION_EVENT` and an identifier for a `rclcpp::SubscriptionBase` object.
+ - `ExecutorEventType::SERVICE_EVENT` and an identifier for a `rclcpp::ServiceBase` object.
+ - `ExecutorEventType::CLIENT_EVENT` and an identifier for a `rclcpp::ClientBase` object.
+ - `ExecutorEventType::WAITABLE_EVENT` and an identifier for a `rclcpp::Waitable` object.
```

Comment:
Thinking out loud, I wonder why combining polymorphism and type tagging, as opposed to just using polymorphism (i.e. making everything a `Waitable`).

Reply:
I agree with that, but it would be a quite big change for ROS as currently the waitable and the subscriptionBase, serviceBase, etc are very distinguished from each other.